### PR TITLE
win32: disable quickedit mode when enabling mouse support

### PR DESCRIPTION
**Pull request by wez**
The Windows conpty layer now allows terminal emulators other
than the classic conhost.exe to host win32 console applications that
use mouse events.

For mouse events to function correctly, mouse input needs to be
enabled and quick edit mode needs to be disabled.

This commit teaches vim to make this change when it enables mouse
mode.

I have no special knowledge of vim, and this patch is roughly based
on the suggested changes in this discussion:

https://groups.google.com/g/vim_dev/c/bQ7jfMwa8Zg

The intent is to enable mouse input and disable quickedit when
enabling mouse mode, and to disable mouse input and restore the
prior quickedit mode state when mouse mode is disabled.

refs: https://github.com/microsoft/terminal/pull/9970

### CODE REVIEW

File: src/os_win32.c [hunk 1]
```diff
@@ -1165,6 +1165,7 @@ decode_key_event(
 
 #endif // FEAT_GUI_MSWIN
 
+static DWORD g_cmodein = 0;
```
Comment (k-takata) [inline src/os_win32.c]: This line should be moved to near line 1184 because this variable is not needed for the GUI.

File: src/os_win32.c [hunk 2]
```diff
@@ -2782,7 +2789,6 @@ SaveConsoleTitleAndIcon(void)
 static int g_fWindInitCalled = FALSE;
 static int g_fTermcapMode = FALSE;
 static CONSOLE_CURSOR_INFO g_cci;
-static DWORD g_cmodein = 0;
 static DWORD g_cmodeout = 0;
```
Comment (k-takata) [inline src/os_win32.c]: I think it's better to move `g_cmodeout` together with `g_cmodein` for readability.

**Comment by k-takata** [comment]
Doesn't the line `cmodein |= g_cmodein & ENABLE_QUICK_EDIT_MODE;` need to be added after here? https://github.com/vim/vim/pull/8280/files#diff-6e09d275741ed4a17372f6e64995645e818d605e41a0591b2c20741f4878d3c1R3778-R3779

**Comment by wez** [comment]
> Doesn't the line `cmodein |= g_cmodein & ENABLE_QUICK_EDIT_MODE;` need to be added after here? https://github.com/vim/vim/pull/8280/files#diff-6e09d275741ed4a17372f6e64995645e818d605e41a0591b2c20741f4878d3c1R3778-R3779

Yeah, good catch! I've restructured this a bit to always set the extended flags when setting the input mode so that it is harder to overlook this again!

**Comment by brammool** [comment]
Can some people try this out and give feedback?

**Comment by wez** [comment]
If someone wants to give it a try, my manual test plan for this is:

* Install either the latest version of `Windows Terminal Preview` (from the MS Store, or download from GitHub) or the Nightly WezTerm build (https://wezfurlong.org/wezterm/install/windows.html)
* Run vim.exe 
* `:set mouse=a`
* Drag the status line or otherwise try to use the mouse for mouse selection of something

Without this change, the mouse won't function.

After exiting vim, run a vim.exe without this change: mouse *shouldn't* work there because quickedit status should have been restored to its default enabled state when vim exited.

**Comment by wez** [comment]
Also: in regular cmd.exe, mouse should work without having to manually change "Quick Edit Mode" in command prompt properties.

**Comment by k-takata** [comment]
I have already used a similar patch based on the old patch in the vim_dev mailing list.
So, I think this should be okay, but I will try this later.

File: src/os_win32.c [hunk 3]
```diff
@@ -1165,7 +1165,6 @@ decode_key_event(
 
 #endif // FEAT_GUI_MSWIN
 
-
```
Comment (k-takata) [inline src/os_win32.c]: This change is unnecessary.

Comment (wez) [reply src/os_win32.c]: fixed!

**Comment by k-takata** [comment]
The following item can be removed from the todo.txt:
```
Win32: Patch for enabling quick edit mode in console. (Craig Barkhouse, 2010
Sep 1)
```

**Comment by brammool** [comment]
Ken Takata wrote:

> The following item can be removed from the todo.txt:
> ```
> Win32: Patch for enabling quick edit mode in console. (Craig Barkhouse, 2010
> Sep 1)
> ```

Didn't we disable quick edit mode with the included change?

-- 
Lose weight, NEVER Diet again with
                  The "Invisible Weight Loss Patch"
						(spam e-mail)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by k-takata** [comment]
Oh, it looks that it's a mistake of the todo item.
Craig's patch actually disabled the QE mode: https://groups.google.com/g/vim_dev/c/bQ7jfMwa8Zg/m/KTEu7WkUAvIJ .

**Pull request closed without merging**